google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,371 @@
+          }
+          timestamp = chunk.getTimestampMicros();
+          labels = chunk.getLabelsList();
+
+          // Update value expectations
+          validate(chunk.getValueSize() >= 0, "NewRow valueSize can't be negative");
+
+          boolean isSplit = chunk.getValueSize() > 0;
+          if (isSplit) {
+            validate(
+                !chunk.getCommitRow(), "NewRow can't commit when valueSize indicates more data");
+            validate(
+                !chunk.getValue().isEmpty(),
+                "NewRow must have data when valueSize promises more data in the next chunk");
+
+            expectedCellSize = chunk.getValueSize();
+            remainingCellBytes = expectedCellSize - chunk.getValue().size();
+          } else {
+            expectedCellSize = chunk.getValue().size();
+            remainingCellBytes = 0;
+          }
+
+          // Start building cell
+          adapter.startCell(familyName, qualifier, timestamp, labels, expectedCellSize);
+          adapter.cellValue(chunk.getValue());
+
+          // Transitions
+          if (isSplit) {
+            return CellInProgress;
+          }
+          adapter.finishCell();
+
+          if (!chunk.getCommitRow()) {
+            return NewCell;
+          }
+
+          return handleCommit();
+        }
+      };
+
+  /** A state that represents a cell's value continuation. */
+  private final State CellInProgress =
+      new State() {
+        @Override
+        State handleChunk(CellChunk chunk) {
+          if (chunk.getResetRow()) {
+            return handleResetChunk(chunk);
+          }
+
+          boolean isLast = chunk.getValueSize() == 0;
+
+          validate(!chunk.hasFamilyName(), "CellInProgress can't have a family");
+          validate(!chunk.hasQualifier(), "CellInProgress can't have a qualifier");
+          validate(chunk.getTimestampMicros() == 0, "CellInProgress can't have a timestamp");
+          validate(chunk.getLabelsCount() == 0, "CellInProgress can't have labels");
+          if (isLast) {
+            long missingBytes = remainingCellBytes - chunk.getValue().size();
+            validate(
+                missingBytes == 0,
+                "CellInProgress is last, but is missing " + missingBytes + " bytes");
+          } else {
+            validate(
+                expectedCellSize == chunk.getValueSize(),
+                "CellInProgress valueSizes should be identical for nonterminal chunks");
+            validate(!chunk.getCommitRow(), "CellInProgress can't commit with a nonterminal chunk");
+          }
+          remainingCellBytes -= chunk.getValue().size();
+          adapter.cellValue(chunk.getValue());
+
+          // Transitions
+          if (!isLast) {
+            return CellInProgress;
+          }
+          adapter.finishCell();
+
+          if (!chunk.getCommitRow()) {
+            return NewCell;
+          }
+
+          return handleCommit();
+        }
+      };
+
+  private static final State CompleteRow =
```

Comment:
"CompleteRow" is super confusing. It looks like a class name. I got confused many times. For example at line#355, we have a local variable "completeRow" and late we return "CompleteRow".

"CompleteRow" is intended to be a state, why can't we name it as such. For example:
READY, PROCESSING_NEW_CELL,  CONTINUE_PROCESSING_CELL, DONE.
These are private static final variables, correct?

We should document all possible states.

Reply:
I'll update the states to be all caps. But I don't agree that the proposed state names are better than the current ones:
- __awaiting new row__: NEW_ROW is more descriptive than READY
- __awaiting new cell__: NEW_CELL makes more sense than PROCESSING_NEW_CELL because haven't start processing a cell yet
- __awaiting more chunks for a cell__: CELL_IN_PROGRESS is about the same as CONTINUE_PROCESSING_CELL
- __finished processing a row, awaiting caller to fetch it__: COMPLETE_ROW makes sense because weren't DONE, we might have more rows to process

Let me know if you are ok with keeping the state names

Reply:
Would this be better?
- AWAITING_NEW_ROW
- AWAITING_NEW_CELL
- AWAITING_CELL_VALUE
- AWAITING_ROW_CONSUME

Reply:
Naming is hard, I suggested something I wasn't happy about.
Your new naming seems to be better.

Reply:
updated with the new names